[analyzer][NFCi] Pass if bind is to a Decl or not to checkBind

gamesh411 opened the pull request:
Binding a value to location can happen when a new value is created or when and existing value is updated. This modification exposes whether the value binding happens at a declaration.
This helps simplify the hacky logic of the BindToImmutable checker.

gamesh411 committed:
[clang][analyzer][NFCi] Pass if bind is to a Decl or not to checkBind

Binding a value to location can happen when a new value is created or
when and existing value is updated. This modification exposes whether
the value binding happens at a declaration.
This helps simplify the hacky logic of the BindToImmutable checker.

gamesh411 commented:
My proposed discussion points about this:
Is the name `atDeclInit` (OK/meaningful enough) to expose to the API users? (I also thought about: isDecl, isInitialization, belongsToDecl, but none of these seemed great)
Does the extra boolean flag introduce any performance overhead? (I'll try to measure this on some open source projects)

steakhal commented on clang/lib/StaticAnalyzer/Checkers/CheckerDocumentation.cpp:
```diff
@@ -175,9 +175,12 @@ class CheckerDocumentation
   /// \param Loc The value of the location (pointer).
   /// \param Val The value which will be stored at the location Loc.
   /// \param S   The bind is performed while processing the statement S.
+  /// \param atDeclInit Whether the bind is performed during declaration
+  ///                  initialization.
```
```suggestion
  /// \param atDeclInit Whether the bind is performed during declaration
  ///                   initialization.
```

steakhal reviewed (commented):
We use UpperCamelCase style for variable names, including parameter names.

Have you checked if any of the checkBind users used an `isa` or `dyncast` as a proxy to determine if its an assignment expression or an initialization?
I suspect there could have been cases where they tried to use this as a proxy, that we could not substitute to this explicit flag instead.

gamesh411 commented:
> Does the extra boolean flag introduce any performance overhead? (I'll try to measure this on some open source projects)
Introducing the flag does make the CheckBindContext struct bigger: https://godbolt.org/z/8PT7br4T3
I will have to run measurements to determine whether this is significant in real use cases.

gamesh411 committed:
[review-fix] rename flag var to AtDeclInit

steakhal reviewed (changes requested):
Ah, just to avoid confusion. Fix the unittests first; then I'll approve :D
EDIT: And also satisfy clang-format.

gamesh411 committed:
fix logic error

calling evalBind was not consistent:
The original issue was that the performTrivialCopy function in
ExprEngineCXX.cpp was always calling evalBind with AtDeclInit = true,
regardless of whether it was handling:
Copy constructors (which should use AtDeclInit = true for
initialization)
Assignment operators (which should use AtDeclInit = false for
assignment)

gamesh411 committed:
fix formatting

gamesh411 commented:
I have stumbled upon a possible inconsistency in the way performTrivialCopy in ExprEngineCXX calls evalBind (by extending the test case to see this new flag as well).
With the above proposed fix, I think it may no longer be NFC technically. Please feel free to remove the tag if you see fit.

steakhal commented:
> With the above proposed fix, I think it may no longer be NFC technically. Please feel free to remove the tag if you see fit.
> Or this fixing of the ExprEngine could become a separate commit, what do you think?

NFCi is fine.

steakhal commented on clang/unittests/StaticAnalyzer/ExprEngineVisitTest.cpp:
```diff
@@ -150,4 +152,23 @@ TEST(ExprEngineVisitTest, checkLocationAndBind) {
+  )",
+                                                    Diags));
+
+  // Look for any occurrence of AtDeclInit = true
+  std::size_t BindPos = Diags.find("AtDeclInit = true");
+  EXPECT_NE(BindPos, std::string::npos);
```
```suggestion
  EXPECT_TRUE(StringRef(Diags).contains("AtDeclInit = true"));
```

gamesh411 committed:
Update clang/unittests/StaticAnalyzer/ExprEngineVisitTest.cpp

Co-authored-by: Balazs Benics <benicsbalazs@gmail.com>

gamesh411 committed:
Update clang/unittests/StaticAnalyzer/ExprEngineVisitTest.cpp

Co-authored-by: Balazs Benics <benicsbalazs@gmail.com>

gamesh411 committed:
Merge branch 'main' into pass-atDeclInit-to-checkBind
